File: co-primitives/src/types/network.rs
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 1io BRANDGUARDIAN GmbH

use crate::CoId;
use co_macros::co;

/// Network service connectivity description.
#[co]
#[non_exhaustive]
pub enum Network {
	/// DID Discovery protocol.
	//#[serde(rename = "/meshsub/1.1.0/diddiscovery/1.0.0")]
	DidDiscovery(NetworkDidDiscovery),

	/// CO Heads protocol.
	//#[serde(rename = "/meshsub/1.1.0/coheads/1.0.0")]
	CoHeads(NetworkCoHeads),

	/// Rendezvouz protocol.
	//#[serde(rename = "/rendezvous/1.0.0")]
	Rendezvous(NetworkRendezvous),

	/// Direct peer connection.
	//#[serde(rename = "/p2p")]
	Peer(NetworkPeer),
}

/// DID Discovery protocol.
#[co]
pub struct NetworkDidDiscovery {
	/// The GossipSub topic used for DidDiscovery messages.
	/// If not specified the default topic will be used: `"co-contact"`.
	pub topic: Option<String>,

	/// The DID to be discovered.
	pub did: String,
}

/// CO Heads protocol.
#[co]
pub struct NetworkCoHeads {
	/// The GossipSub topic used for Heads messages.
	/// If not specified the default topic will be used: `"co-{co.id}"`.
	pub topic: Option<String>,

	/// The CO to be discovered.
	pub id: CoId,
}

/// Rendezvouz protocol.
#[co]
pub struct NetworkRendezvous {
	/// The namespace to register to.
	pub namespace: String,
	/// Rendezvouz node multi-addresses.
	pub addresses: Vec<String>,
}

/// Direct peer connection.
#[co]
pub struct NetworkPeer {
	/// The [`libp2p::PeerId`] as bytes.
	pub peer: Vec<u8>,
	/// Optional known Multi-addresses.
	/// If empty, mDNS, Kademila and Bluetooth LE may be used to resolve.
	pub addresses: Vec<String>,
}

File: tauri-plugin-co-sdk/guest-js/types/network.ts
import { CoId } from "./co.js";

/** Enum (serde external-tag) */
export type Network =
  | { DidDiscovery: NetworkDidDiscovery }
  | { CoHeads: NetworkCoHeads }
  | { Rendezvous: NetworkRendezvous }
  | { Peer: NetworkPeer };

/** DID Discovery protocol. */
export interface NetworkDidDiscovery {
  /** The GossipSub topic used for DidDiscovery messages. Default: "co-contact". */
  topic?: string; // Option<String>
  /** The DID to be discovered. */
  did: string;
}

/** CO Heads protocol. */
export interface NetworkCoHeads {
  /** The GossipSub topic used for Heads messages. Default: "co-{co.id}". */
  topic?: string; // Option<String>
  /** The CO to be discovered. */
  id: CoId;
}

/** Rendezvous protocol. */
export interface NetworkRendezvous {
  /** The namespace to register to. */
  namespace: string;
  /** Rendezvous node multi-addresses. */
  addresses: string[];
}

/** Direct peer connection. */
export interface NetworkPeer {
  /** libp2p PeerId as bytes. (serde_json: number[] unless using serde_bytes for base64) */
  peer: number[]; // Vec<u8>
  /** Optional known multi-addresses. */
  addresses: string[];
}
